apps/src/lib/util/urlHelpers.js

```diff
@@ -13,3 +15,37 @@ export function pegasus(relativeUrl) {
   }
   return relativeUrl;
 }
+
+/**
+ * Retrieve the meta description tag from the specified url
+ * Memoize so that we only request once per relative url.
+ */
+export function metaTagDescription(relativeUrl) {
+  return metaTagMemoized(relativeUrl);
+}
+
+/**
+* Fetch the html of the given page and parse to get content of
+* 'description' meta tag.
+*/
+const metaTagMemoized = _.memoize((relativeUrl) => {
+  return fetch(relativeUrl)
+    .then(
+      response => {
+        // Catch fetch's 400 errors
+        if (response.status < 200 || response.status >= 300) {
+          return relativeUrl;
+        } else {
+          return response.text().then(data => {
+            const metaTag = $(data).filter("meta[name='description']").attr("content");
+            // Return url if there was no description meta tag
+            if (metaTag) {
+              return metaTag;
+            } else {
+              return relativeUrl;
+            }
+          });
+        }
+      }
+    ).catch(error => relativeUrl);
+});
```

Comment:
I'm convinced this works great, but there are some style things, especially with promise chains, that could be cleaned up.

In general, promise chains inside promise chains is a code smell, and might indicate you should break the work up differently.  Here's what I'd do:

```js
return fetch(relativeUrl)
  .then(response => Promise.all([response.status, response.text()]))
  .then((status, text) => {
    if (status < 200 || status >= 300) {
      return relativeUrl;
    } else {
      const metaTag = $(text).filter("meta[name='description']").attr('content');
      return metaTag || relativeUrl;
    }
  })
  .catch(() => relativeUrl);
```
The first `.then` clause does a couple of things - it does additional async work, and it only passes along the parts of the response we care about.  This is slightly less optimal than your solution (you're only doing the async work of reading the response body if the status is OK) but I think it's easier to follow.

Reply:
This is beautiful! I knew there had to be a way to make this promise chain readable and prettier. Thanks for the assist!